ahcore/callbacks/abstract_writer_callback.py

```diff
@@ -333,6 +367,31 @@ def _monitor_and_cleanup(self) -> None:
             time.sleep(0.5)  # Short sleep to yield control and reduce CPU usage
 
     def _epoch_end(self, trainer: pl.Trainer, pl_module: pl.LightningModule) -> None:
+        for callback in self._callbacks:
+            while self._num_filenames_seen != callback.completed_tasks:
```

Comment:
During Sanity Checking with two gpus, I get the following error in one of the processes. The main thread will forever (?) hang here. This happened before training of an AttentionUnet. I did not encounter this problem when switching to training on a single gpu. See full traceback below.

---
[2024-05-07 23:21:21,052][ahcore.callbacks.converters.common][INFO] - Starting worker for ComputeWsiMetricsCallback
Process Process-7:
Traceback (most recent call last):
  File "/home/ba.d.rooij/miniconda3/envs/dev_ahcore/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/home/ba.d.rooij/miniconda3/envs/dev_ahcore/lib/python3.10/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/ba.d.rooij/dev_ahcore/ahcore/ahcore/callbacks/converters/common.py", line 77, in worker
    result = self.process_task(filename, cache_filename)
  File "/home/ba.d.rooij/dev_ahcore/ahcore/ahcore/callbacks/converters/wsi_metric_callback.py", line 103, in process_task
    curr_metrics = compute_metrics_for_case(
  File "/home/ba.d.rooij/dev_ahcore/ahcore/ahcore/callbacks/converters/wsi_metric_callback.py", line 172, in compute_metrics_for_case
    wsi_metrics.process_batch(
  File "/home/ba.d.rooij/dev_ahcore/ahcore/ahcore/metrics/metrics.py", line 352, in process_batch
    metric.process_batch(predictions, target, wsi_name=wsi_name, roi=roi)
  File "/home/ba.d.rooij/dev_ahcore/ahcore/ahcore/metrics/metrics.py", line 196, in process_batch
    dice_components = _get_intersection_and_cardinality(
  File "/home/ba.d.rooij/dev_ahcore/ahcore/ahcore/metrics/metrics.py", line 381, in _get_intersection_and_cardinality
    predictions = soft_predictions.argmax(dim=1)
RuntimeError: pool INTERNAL ASSERT FAILED at "../aten/src/ATen/ParallelOpenMP.cpp":65, please report a bug to PyTorch. Invalid thread pool!
Sanity Checking ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 971/971 0:07:44 • 0:00:00 3.29it/s  [2024-05-07 23:28:59,863][ahcore.callbacks.abstract_writer_callback][INFO] - Validation epoch ended.

Reply:
I think this might be the OpenMP num threads. Can you try setting environment variable "OMP_NUM_THREADS" equal to 1, and try running again? If that fixes it, we can include a warning / set it automatically perhaps.

Reply:
Setting OMP_NUM_THREADS equal to 1 fixes the problem. The RuntimeError does not get thrown, the sanity check completes successfully and training starts as it should.